### PR TITLE
C++: Refactor ExecTainted.ql to only report results after string concatenation

### PR DESCRIPTION
**Pull request by rdmarsh2**
This makes ExecTainted report results only when the tainted value does not become the start of the string which is eventually run as a shell command. The theory is that those cases are likely to be deliberate, and part of the expected threat model of the program (e.g. $CC in make). This lines up better with the results I considered fixable true positives in LGTM testing.

Opened as a draft because I have more work to do on testing, and it's in need of a rebase to clean up old commits and give before/after results for the newly added tests

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:42]: You will also occasionally see `char16_t *` strings, e.g. in ChakraCore.  Could we widen this to any pointer to an `IntegralType` (or perhaps look at the `getConversionChar` instead)?

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:33]: This predicate badly needs an explanation of what it's doing, especially the `not literal.getConvSpecOffset(index) = 0` line.  I can see from the context of this PR and the commit comment, but future maintainers won't (easily) be able to see that.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:67]: If I understand correctly, this strategy ensures that taint passes through at least one 'interesting' concatenation.  Two details spring to mind:
 - if there are no concatenations or string processing at all, just an `exec` of the tainted string as it is, currently I think this will not be reported.  Strictly speaking the tainted string *is* at the start of the `exec` string, but is this what we want to do?
 - we no longer treat `printf("%i", taint)` as a barrier, merely as a non-interesting concatenation, which I think will fail to rule out some multi-step cases that we'd have ruled out when it was a barrier.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:59]:
```suggestion
    call.getArgument(1) = rse.getArgumentOperand().getAnyDef() and // right operand
```

**Review by geoffw0 — COMMENTED** [review]
This looks promising but a few more changes are needed.  After that I'd really like to see how results on LGTM will be affected.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:42]: Good point - I think `getConversionChar` is the right way to go.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:58]: Whenever I see code like this I feel like the decision to make all of the models `private` was a mistake. IMO, I'd much rather want to write `op instanceof StdStringPlus` rather than this somewhat ugly (but necessary) conjunction.

Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:58]: I agree, but that's something to take up elsewhere.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:67]: >  Strictly speaking the tainted string is at the start of the exec string, but is this what we want to do?

My assumption was that it would be a deliberate decision (like `codeql databse create -c`).

> we no longer treat printf("%i", taint) as a barrier

Good point. Maybe a barrier on any value with an integral type.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:40]: Added

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:70]: Added

**Comment by rdmarsh2** [comment]
rebased and responded to PR comments. New LGTM run: https://lgtm.com/query/1936393694543588922/

**Review by geoffw0 — COMMENTED** [review]
QL and results otherwise LGTM.

I note the query now produces much fewer (and different) results to before.  They appear to be high quality results, which is what we want.

**Comment by rdmarsh2** [comment]
I've added that QLDoc, raised the precision to `high`, and added a change note. Let me know if you want a bigger LGTM run or a medium precision.

**Comment by geoffw0** [comment]
I'm happy with results, a CPP-Differences might be good, to confirm performance.

File: cpp/ql/src/semmle/code/cpp/commons/Printf.qll
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/commons/Printf.qll]: Thanks.

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:118]: Re: the TODO.  I don't really see how.  Did you have anything in mind?

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:88]: Shouldn't this `TaintTracking::Configuration` be `TaintTracking2::Configuration`?

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:118]: I was concerned about losing call contexts when we switch data flow configurations. I didn't see any false positives caused by it, and I don't think the dataflow API exposes call contexts on PathNode. If we get flow labels in the future, they'd be a cleaner solution to this problem.

**Comment by rdmarsh2** [comment]
Started a CPP-Differences run: [internal link](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2169/)

**Comment by rdmarsh2** [comment]
rerunning the CPP-Differences: [internal link](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2174/)

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:88]: Yeah, for performance. Unfortunately that breaks the combined paths I'd set up - I might be able to make it work with an IPA type but I'm not confident about it.

**Comment by geoffw0** [comment]
Changes LGTM.  Is performance OK?

**Comment by rdmarsh2** [comment]
The last rerun [here](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2178/) has slowdowns of 17% to 60%, all attributed to ExecTainted.ql. I think there's reevaluation going on, either of the IR, dataflow, or both.

**Comment by MathiasVP** [comment]
> The last rerun [here](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2178/) has slowdowns of 17% to 60%, all attributed to ExecTainted.ql. I think there's reevaluation going on, either of the IR, dataflow, or both.

Looking at the profile viewer, the IR is definitely being re-evaluated, yeah.

**Comment by rdmarsh2** [comment]
> I would guess that a fix similar to #6347 would solve this.

So would I, but I haven't managed to find the right import order...

File: cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:87]: Would it be better for performance if we implemented this `isSource` like this:
```codeql
override predicate isSource(DataFlow::Node source) {
  exists(DataFlow::Node prevSink, TaintToConcatenationConfiguration conf |
    conf.hasFlow(_, prevSink) and
    interestingConcatenation(prevSink, source)
  )
}
```
so that we only compute flow from the sources that actually end up having flow from a `FlowSource`?

Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql:87]: Good call.

**Comment by rdmarsh2** [comment]
Rebased to pick up #5032 and add forwarding for the new `PathGraph::subpaths` predicate, and started a new [CPP-Differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2244/)

**Comment by geoffw0** [comment]
Code changes LGTM.

CPP Differences LGTM.  All differences appear to be in the `BAD` cases for `CWE78_OS_Command_Injection`, which is exactly what we want.

Checks are failing.  I think these are due to the (now unnecessary) import of `import semmle.code.cpp.security.FunctionWithWrappers` in `System.qll`, bringing in the `PrintableElement` class from `PrintAST.qll`.

Performance - do we trust CPP-Differences for this or should I check locally?

**Comment by MathiasVP** [comment]
> Performance - do we trust CPP-Differences for this or should I check locally?

If you're referring to https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2244/ then indeed there is/was a problem. I think this is what Robert wrote about on Slack here:
> ExecTainted... turns out semmle.code.cpp.security.CommandExecution adds to models that the IR uses and actually changes the semantics. I'll move it into the main models library tomorrow.

I think this has been fixed by a3e1f54e3309a4b1c0602110b886ecd69941f8d0, but we should run a fresh CPP-Differences to confirm this.

**Comment by geoffw0** [comment]
Recent changes LGTM.  `ResolveCall` looks like something we could begin to use more widely?

**Comment by MathiasVP** [comment]
> Recent changes LGTM. `ResolveCall` looks like something we could begin to use more widely?

Indeed. We've previously received a couple of requests to support function pointer resolution outside the context of global dataflow. Now we can actually point people to a public pretty interface instead of internal dataflow files that manipulate the IR directly 🎉.

**Comment by jbj** [comment]
All checks are passing! Is this PR ready for final review? If so, @rdmarsh2, please share links to
- A DCA run to demonstrate that performance and results are alright on our set of test projects.
- A test run of the query (with its supporting libraries if possible) on our sample of ~100 projects on lgtm.com. See `Adding a new query.md` in our team repo.

**Comment by rdmarsh2** [comment]
Yes, it is.

DCA run: [https://github.com/github/codeql-dca-main/issues/508](internal link)
LGTM run: https://lgtm.com/query/4837668389671482937/

**Comment by rdmarsh2** [comment]
> I was concerned about losing call contexts when we switch data flow configurations. I didn't see any false positives caused by it, and I don't think the dataflow API exposes call contexts on PathNode. If we get flow labels in the future, they'd be a cleaner solution to this problem.

The result on WohlSoft/Moondust-Project in the latest LGTM run is a false positive caused by this...

**Comment by jbj** [comment]
I'll also point out (and this is totally not a matter of vanity :innocent:) that the result on jbj/magicrescue is a FP. I see you've added `test7`, which demonstrates the same kind of FP.

**Review by jbj — APPROVED** [review]

**Pull request merged**
